### ISSUE TITLE
Incorrect parsing of "meta" box for screen recording file with Quicktime

### ISSUE THREAD

**Issue by vjeux** [gpac/mp4box.js#309]
I saved a screen recording mp4 file with sound with Quicktime ([mp4_with_sound.mov.zip](https://github.com/gpac/mp4box.js/files/10346002/mp4_with_sound.mov.zip)). If you open this file in the [filereader.html](https://gpac.github.io/mp4box.js/test/filereader.html), you can see this in the console:

```
[BoxParser] 'meta' box writing not yet implemented, keeping unparsed data in memory for later write
[BoxParser] Box of type '' has a size 1751411826 greater than its container size 134
[BoxParser] Parsing of box 'meta' did not read the entire indicated box data size (missing 126 bytes), seeking forward
```

**Comment by selsamman** [comment]
I am getting a similar error with any video shot on the Samsung Galaxy S10.  In my case as far as I can tell segmenting still works and the moov box is found and seems reasonable.  I am able to stream the segments over WebRTC to MSE in a browser.  And BTW this is an absolutely awesome package.  Using the latest version of mp4box.js

[0:00:00.001] [ISOFile] Processing buffer (fileStart: 0)
[0:00:00.002] [BoxParser] Unknown box type: 'SDLN'
[0:00:00.003] [BoxParser] 'SDLN' box writing not yet implemented, keeping unparsed data in memory for later write
[0:00:00.003] [BoxParser] Unknown box type: 'smrd'
[0:00:00.003] [BoxParser] 'smrd' box writing not yet implemented, keeping unparsed data in memory for later write
[0:00:00.003] [BoxParser] Unknown box type: '©xyz'
[0:00:00.003] [BoxParser] '©xyz' box writing not yet implemented, keeping unparsed data in memory for later write
[0:00:00.003] [BoxParser] Unknown box type: 'smta'
[0:00:00.003] [BoxParser] 'smta' box writing not yet implemented, keeping unparsed data in memory for later write
[0:00:00.003] [BoxParser] 'meta' box writing not yet implemented, keeping unparsed data in memory for later write
[0:00:00.003] [BoxParser] Box of type '' has a size 1751411826 greater than its container size 165
[0:00:00.003] [BoxParser] Parsing of box 'meta' did not read the entire indicated box data size (missing 157 bytes), seeking forward
[0:00:00.005] [BoxParser] 'colr' box writing not yet implemented, keeping unparsed data in memory for later write
[0:00:00.006] [BoxParser] 'esds' box writing not yet implemented, keeping unparsed data in memory for later write
[0:00:00.008] [ISOFile] Done processing buffer (fileStart: 0) - next buffer to fetch should have a fileStart position of 5944036
[0:00:00.008] [MultiBufferStream] 1 stored buffer(s) (3436/5944036 bytes), continuous ranges: [0-5944035]
[0:00:00.008] [ISOFile] Sample data size in memory: 0
[0:00:00.008] [ISOFile] Flushing remaining samples
[0:00:00.008] [MultiBufferStream] 1 stored buffer(s) (3436/5944036 bytes), continuous ranges: [0-5944035]

**Comment by vjeux** [comment]
@davemevans great investigation work! Is there a way to distinguish which version we are using? If not, we could try to parse using one format and if it doesn't fit the number of bytes try parsing using the second format. It's not ideal but would work.

**Comment by bradh** [comment]
We could just bail after the `ftyp` box if we find a  major brand for quicktime.

**Comment by tobiasBora** [comment]
Same error here for video taken with a Samsung XCover:
```
[0:00:11.885] [BoxParser] Box of type '' has a size 1751411826 greater than its container size 165
```

**Comment by jozefchutka** [comment]
I have a same error with video taken with Xiaomi phone
```
[0:00:00.009] [BoxParser] Box of type '' has a size 1751411826 greater than its container size 225
```

```
General
Format                                   : MPEG-4
Format profile                           : Base Media / Version 2
Codec ID                                 : mp42 (isom/mp42)
File size                                : 77.4 MiB
Duration                                 : 31 s 667 ms
Overall bit rate                         : 20.5 Mb/s
Frame rate                               : 30.000 FPS
Encoded date                             : 2023-08-30 10:22:20 UTC
Tagged date                              : 2023-08-30 10:22:20 UTC
xyz                                      : +35.3654+034.0776/
com.android.version                      : 13
com.android.manufacturer                 : Xiaomi
com.android.model                        : M2101K6G

Video
ID                                       : 1
Format                                   : AVC
Format/Info                              : Advanced Video Codec
Format profile                           : High@L1
Format settings                          : CABAC / 2 Ref Frames
Format settings, CABAC                   : Yes
Format settings, Reference frames        : 2 frames
Codec ID                                 : avc1
Codec ID/Info                            : Advanced Video Coding
Duration                                 : 31 s 667 ms
Source duration                          : 31 s 666 ms
Bit rate                                 : 20.0 Mb/s
Width                                    : 1 920 pixels
Height                                   : 1 080 pixels
Display aspect ratio                     : 16:9
Frame rate mode                          : Variable
Frame rate                               : 30.000 FPS
Minimum frame rate                       : 14.928 FPS
Maximum frame rate                       : 10 000.000 FPS
Standard                                 : NTSC
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 8 bits
Scan type                                : Progressive
Bits/(Pixel*Frame)                       : 0.322
Stream size                              : 75.6 MiB (98%)
Source stream size                       : 75.6 MiB (98%)
Title                                    : VideoHandle
Language                                 : English
Encoded date                             : 2023-08-30 10:22:20 UTC
Tagged date                              : 2023-08-30 10:22:20 UTC
Color range                              : Limited
colour_range_Original                    : Full
Color primaries                          : BT.709
colour_primaries_Original                : BT.601 PAL
Transfer characteristics                 : BT.709
transfer_characteristics_Original        : BT.601
Matrix coefficients                      : BT.709
matrix_coefficients_Original             : BT.470 System B/G
mdhd_Duration                            : 31667
Codec configuration box                  : avcC

Audio
ID                                       : 2
Format                                   : AAC LC
Format/Info                              : Advanced Audio Codec Low Complexity
Codec ID                                 : 2 / 40 / mp4a-40-2
Duration                                 : 31 s 592 ms
Source duration                          : 31 s 573 ms
Bit rate mode                            : Constant
Nominal bit rate                         : 96.0 kb/s
Channel(s)                               : 2 channels
Channel layout                           : L R
Sampling rate                            : 48.0 kHz
Frame rate                               : 46.875 FPS (1024 SPF)
Compression mode                         : Lossy
Source stream size                       : 370 KiB (0%)
Title                                    : SoundHandle
Language                                 : English
Encoded date                             : 2023-08-30 10:22:20 UTC
Tagged date                              : 2023-08-30 10:22:20 UTC
mdhd_Duration                            : 31592
```

**Comment by RufaelDev** [comment]
Yes I observed the same issue as @davemevans 2 years ago filming on my mobile device (huawei p30 lite), many android phones still output quicktime type metabox, but files do not include the brand in the ftyp box.

**Comment by RufaelDev** [comment]
14496-12 defines meta as FullBox('meta', version = 0, 0), so maybe the best way to parse meta to look at the byte following `meta`, if it is zero it is an isobmff box, if it is not it is a new box enclosed in the quicktime meta box  (non full box) ? does it make sense ? I do not think the brand in ftyp will work as it is usually not set.

**Comment by bradh** [comment]
> 14496-12 defines meta as FullBox('meta', version = 0, 0), so maybe the best way to parse meta to look at the byte following `meta`, if it is zero it is an isobmff box, if it is not it is a new box enclosed in the quicktime meta box (non full box) ? does it make sense ? I do not think the brand in ftyp will work as it is usually not set.

I'm not sure that will work. What is the byte following `meta` in your quicktime sample? If its another box, it'll start with a four byte length, and the most likely first byte in that length is 0.

**Comment by RufaelDev** [comment]
@bradh you need to check 4 bytes, it will be zero for fullbox (isobmff) and non zero for quicktime box (hdlr)

**Comment by bradh** [comment]
Possible. Does quicktime use 0x00 0x00 0x00 0x00  in the same way as the ISO BMFF  - to mean "to end of file"?

**Comment by bradh** [comment]
https://issuetracker.google.com/issues/232971800 indicates that Google are going to keep doing this. They suggest a slightly different work around to parsing based on checking for `hdlr` 4cc.

**Comment by stefwalter** [comment]
I've started a set of extensions to MP4Box.js which solves this problem for me: https://github.com/stefwalter/qtffbox.js

Am I correct in assuming that such QuickTime support is outside of MP4Box.js's scope?

**Comment by stefwalter** [comment]
> > Am I correct in assuming that such QuickTime support is outside of MP4Box.js's scope?
> 
> I'd say it is in scope since the main GPAC support QT. As a consequence community contributions are always welcome :)

Awesome. 

> Do you feel your code should be integrated in this repo (preferred option for maintenance)? Or should we reference it from the doc?

Yes, I'd be happy for integrating into this repo. How about I work on a pull request and we can discuss specifics there?

**Comment by DenizUgur** [comment]
I agree, if you open a PR I'll be happy to go over it.

**Comment by stefwalter** [comment]
> I agree, if you open a PR I'll be happy to go over it.

Thanks. I'm working on it. But stuck on https://github.com/gpac/mp4box.js/issues/357 before I can contribute.

**Comment by DenizUgur** [comment]
The conformance files should be accessible from here:

https://github.com/MPEGGroup/FileFormatConformance/tree/main/data%2Ffile_features%2Fpublished%2Fisobmff

The repository is due for an update on the tests but in the meantime don't hesitate to open the PR for a review.
